Check for ANSI colors in all plain text outputs

mgeier opened the pull request:
This is yet another alternative to #348.

It makes somewhat less changes, in particular it keeps the separation between "fancy" and "non-fancy" output cells (with regards to LaTeX).

But like #348, it removes the separation between "ANSI" and "text/plain" outputs.

Here's a PDF preview: https://51-210404706-gh.circle-artifacts.com/0/nbsphinx.pdf#section.4

jfbu commented:
It seems you keep the LaTeX changes to a minimum, and it solves the problem on your side, then this looks good! For info I compiled nbsphinx.pdf in this branch using both Sphinx current 2.0 branch as well as my [WIP branch](https://github.com/sphinx-doc/sphinx/compare/2.0...jfbu:6804_long_code_block_in_danger) (which I will turn into a PR right away) and I detected no problem with the enhanced `sphinxVerbatim` the WIP branch promotes to fix sphinx-doc/sphinx#6804. However I examined PDF very quickly.

mgeier commented:
Thanks a lot @jfbu! I've also tried it with your branch and haven't found any problems.

I've found an unrelated problem, though: The single quotes are different in input and output cells:

![image](https://user-images.githubusercontent.com/705404/68544564-cc0b7e00-03c4-11ea-9740-62db2a078314.png)

See: https://51-210404706-gh.circle-artifacts.com/0/nbsphinx.pdf#subsection.7.4

It looks like Pygments replaces the quotes with `\PYGZsq{}`.

Do you happen to know how I can also use straight quotes in the output cell?

Do I have to use `\sphinxupquote{}` or is there a setting to apply this to the whole `sphinxVerbatim` environment (or globally to all environments)?

jfbu commented:
Hmm... if I copy over the

```latex
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{red}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}\PYG{l+s+s1}{\PYGZsq{}}\PYG{l+s+s1}{no problem}\PYG{l+s+s1}{\PYGZsq{}}
\end{sphinxVerbatim}
```

(changing color) into the output of a sphinx project I had handy (hence the background color) I have no problem in output. 

![Capture d’écran 2019-11-10 à 14 36 37](https://user-images.githubusercontent.com/2589111/68544859-87351680-03c7-11ea-8252-6ea3cda29fe9.png)

I need to understand what's going wrong. But I was going to leave, so if you don't hear from me in next half-hour I will be offline until tonight European time.

Sphinx does take car to handle the `\PYGZsq` in its `\sphinxbreaksatspecials`.

s-weigand commented:
The HTML output in [my usecase](https://github.com/students-teach-students/python-tools-for-students) (compare with #330 ), also looks as it should now. 😃 
Thanks @mgeier 

**New HTML render:**
![npsphinx_unicode_fix](https://user-images.githubusercontent.com/9513634/68544933-412c8280-03c8-11ea-84a2-bbeb3771b4dd.jpg)

jfbu commented:
I am sooo stupid. Of course the relevant part of nbsphinx.tex is

```
\begin{sphinxVerbatim}[commandchars=\\\{\}]
\llap{\color{nbsphinxout}[1]:\,\hspace{\fboxrule}\hspace{\fboxsep}}'no problem'
\end{sphinxVerbatim}
```

there the ``'`` is naked and not escaped to `\PYGZsq{}`.

jfbu commented:
a bit short on time but the story is that `sphinxVerbatim` does not make `'` an active character because it is supposed to have been tex-escaped by Pygmentize library. It does

```
\let\verbatim@nolig@list \sphinx@verbatim@nolig@list
```

and one has a very short  list ``\def\sphinx@verbatim@nolig@list {\do \`}``, because all others are TeX-escaped, execpt the comma but the comma is handled especially to allow linebreaks and Sphinx does not want `fancyvrb` to override this. Thus the very short `\verbatim@nolig@list` is what gets executed by `fancyvrb` via `\@noligs` and it thus misses making `'` an active character (which is needed for it to avoid being rendered curly because even monospace TeX fonts usually have it curly).

Thus

```latex
{\catcode`\'\active
\begin{sphinxVerbatim}[commandchars=\\\{\}]
'no problem'
\end{sphinxVerbatim}
}
```

fixes it.

jfbu commented:
Maybe slightly better way is 

```latex
{\fvset{codes*=\catcode`\'\active}
\begin{sphinxVerbatim}[commandchars=\\\{\}]
'no problem'
\end{sphinxVerbatim}
}
```

and probably you can issue only once in preamble

```
\fvset{codes*=\catcode`\'\active}
```

I don't expect bad things out of adding this to preamble. The effet is to add activation of `'` in (sphinx)Verbatim environments. It is done by default in original Verbatim, but not in sphinxVerbatim for reasons explained above of not being needed a priori.

jfbu commented:
> It looks like Pygments replaces the quotes with `\PYGZsq{}`.
> 

Yes

> Do you happen to know how I can also use straight quotes in the output cell?
> 

I had misunderstood your question initially but the answer is that the `'` character must be made "active". It is done by original `fancyvrb`'s  `Verbatim`  but not by `sphinxVerbatim` because the latter does not need it (cf `PYGZsq`) and must override some of `fancyvrb`'s  set-up and it did that in a way not handling the `'` character (as not being needed).

> Do I have to use `\sphinxupquote{}` or is there a setting to apply this to the whole `sphinxVerbatim` environment (or globally to all environments)?

The ``\fvset{codes*=\catcode`\'\active)`` is the way to add the config to all `sphinxVerbatim` environments.

Better to avoid making the `'` active globally because its active definition is the one for math mode and only verbatim like environments (which execute `\@noligs`)  confer to it the `\textquotesingle` meaning adapted for text mode.

mgeier commented:
That works perfectly, thanks @jfbu!
I've added your suggestion in edebf63078f6d4f0df6cf18177179d0358b73bd5.

Now the quotes are straight: https://53-210404706-gh.circle-artifacts.com/0/nbsphinx.pdf#subsection.7.4

@s-weigand Thanks for testing!

I will leave this PR open for a few days and if nothing else comes up, I will merge it (and close the other alternatives).
And then I think I'll make a new release.

jfbu commented:
If I get this right you are using `sphinxVerbatim` with contents not having been malaxed by Pygmentize? Thus for example we can very well have  `--`  in the LaTeX cell code? Basically, things will be as TeX-escaped as done by https://github.com/sphinx-doc/sphinx/blob/2.0/sphinx/util/texescape.py, right? Then we need to be a bit more pro-active to avoid mechanism of TeX ligatures transforming `--` or `---`.

Why has `sphinxVerbatim` no provision for that? because it assumed the Pygments mark-up is used (`\PYGZhy`). The fix is relatively simple. Add to LaTeX preamble:

```
\makeatletter
% Sphinx default only does \do\` because other characters are assumed Pygmentized
% the ' and ` will remain straight in PDF output
% the - will get prevented from forming TeX ligatures in PDF output
\def\sphinx@verbatim@nolig@list{\do\'do\`\do\-}
\makeatother
```

As the above handles the straight quote `'` the `\fvset` become superfluous and can be dropped. The above is to be inserted in preamble (after `sphinx` package loading).

A priori the `<` and `>`  can give TeX ligatures too but they need no handling as I expect them escaped to `r'\textless{}'` and ` r'\textgreater{}'`, if I understand correctly @mgeier context here. In future, btw, Sphinx will probably use rather `r'\sphinxtextless{}'` and ` r'\sphinxtextgreater{}'` to allow special customizations. There is currently code in `sphinx.sty` for "parsed-literals" which will then need to adjust to the new names.

See `\sphinxbreaksattexescapedchars` in `sphinx.sty`; it could make sense to use that macro here for those `sphinxVerbatim` environment which are used with non-Pygmentized contents: then soft wrap of long lines will work about the same as in normal `sphinxVerbatim` having only Pygmentized contents.

I am sorry that this comment indicates something to do and then goes to say perhaps even something else needs to be done, but I have no clear understanding of what exactly can end up inside these `sphinxVerbatim` blocks.

There is also `\sphinxbreaksviaactiveinparsedliteral` which could be added and it handles the `-` so the `-` would then need to be removed again from above hint at top of this comment.

The syntax would be something like 

```
\fvset{codes*=\catcode`\'\active\sphinxbreaksattexescapedchars\sphinxbreaksviaactiveinparsedliteral}
```

I have tested it very briefly. I *think* it can be done globally in preamble, i.e. that the impact on standard code cells using `sphinxVerbatim` with Pygmentized contents is nil, but this could require testing.

What you gain with the above is that long lines will wrap at those characters.

Beware however if you end up with math mode things in such cells? Is it at all possible? Sphinx code for parsed-literals has extra steps for those.

Can you have a `\url` inside such environments?

mgeier commented:
> If I get this right you are using `sphinxVerbatim` with contents not having been malaxed by Pygmentize?

Yes.
Because I want to apply ANSI control sequences, and I don't think that's possible with Pygments, right?

In fact, I could use a `docutils.nodes.literal_block` and afterwards apply the `ansi2latex` filter, but this would only work for LaTeX output. I didn't find a way to make this work with HTML output.

That's why I'm not using `docutils.nodes.literal_block` but doing it manually with `raw` directives.

> Thus for example we can very well have  `--`  in the LaTeX cell code?

Yes, totally.
Anything that can be created by Pythons `print()` function.
This is typically plain text, and it can contain control characters. Especially ANSI escape sequences.

> Basically, things will be as TeX-escaped as done by https://github.com/sphinx-doc/sphinx/blob/2.0/sphinx/util/texescape.py, right?

Currently the content of code cell outputs is passed through the filters `escape_latex` and `ansi2latex`, which are both provided by `nbconvert`:

https://github.com/jupyter/nbconvert/blob/b0209dfd029da8797a144023b008198c3e36aa32/nbconvert/filters/latex.py#L30-L64

https://github.com/jupyter/nbconvert/blob/b0209dfd029da8797a144023b008198c3e36aa32/nbconvert/filters/ansi.py#L64-L74

`escape_latex` is apparently doing fewer replacement than `texescape.py`, but I guess that's an `nbconvert` issue.

> ```
> \makeatletter
> \def\sphinx@verbatim@nolig@list{\do\'do\`\do\-}
> \makeatother
> ```

This doesn't seem to work.

I don't know why exactly, but I get an error from LaTeX:

```
! Improper alphabetic constant.
<to be read again> 
                   \global 
l.325 \begin{sphinxVerbatim}[commandchars=\\\{\}]
                                                 
? 
```

> A priori the `<` and `>`  can give TeX ligatures too but they need no handling as I expect them escaped to `r'\textless{}'` and ` r'\textgreater{}'`, if I understand correctly @mgeier context here.

No,  `<` and `>` are not replaced and e.g. `>>` becomes a ligature `»`!

> ```
> \fvset{codes*=\catcode`\'\active\sphinxbreaksattexescapedchars\sphinxbreaksviaactiveinparsedliteral}
> ```

This seems to work well!
(except for `<` and `>`)

> Beware however if you end up with math mode things in such cells? Is it at all possible?

No. This is only for plain text.
Math is handled separately in "fancy" outputs.

> Can you have a `\url` inside such environments?

No, actual LaTeX output is also handled in "fancy" outputs.
If there is LaTeX markup in "non-fancy" outputs, it should be shown verbatim in plain text.

This is how math/LaTeX output is supposed to look like (using "fancy" outputs): https://buildmedia.readthedocs.org/media/pdf/nbsphinx/0.4.3/nbsphinx.pdf#subsubsection.4.3.3

jfbu commented:
> > ```
> > \makeatletter
> > \def\sphinx@verbatim@nolig@list{\do\'do\`\do\-}
> > \makeatother
> > ```
> 
> 
> This doesn't seem to work.

Very sorry about typo (missing backslash in second `do`) it should have read

```latex
\makeatletter
\def\sphinx@verbatim@nolig@list{\do\'\do\`\do\-}
\makeatother
```

(but read further down)

> No, `<` and `>` are not replaced and e.g. `>>` becomes a ligature `»`!
> 
> > ```
> > \fvset{codes*=\catcode`\'\active\sphinxbreaksattexescapedchars\sphinxbreaksviaactiveinparsedliteral}
> > ```
> 
> This seems to work well!
> (except for `<` and `>`)

Taking into account the information you provided I can propose the following possibly not quite optimized:

```
\makeatletter
\renewcommand*\sphinx@verbatim@nolig@list{\do\'\do\`}
\makeatother
\renewcommand*\sphinxbreaksbeforeactivelist{\do\<}
\renewcommand*\sphinxbreaksafteractivelist{\do\.\do\,\do\:\do\;\do\?\do\!\do\/\do\>\do\-}
\fvset{codes*=\sphinxbreaksattexescapedchars}
```

I was worried about the hyphen character `-` when I saw mark-up such as ``\colorbox{ansi-yellow}{CD\strut}`` but examining a TeX trace I discovered that `xcolor` package has some sanitizing of its own when certain characters have been made active, and in particular the hyphen `-` as present here in `ansi-yellow` name.

The characters sanitized by `xcolor` in such color names seem to be `!:-+;">` but for some reason not the `<`. So you can use `>` in a color name (not that I recommend it...) but not a `<` is such color names end up in  environments "activating" `<`, `>` like is the case  of `sphinxVerbatim` with above config.

In the above proposed config, the single quotes ``'` `` are only make active to become straight quotes, but will not offer potential breakpoints. For this my impression is that a specific macro is to be added in place of simply like in the above recycling some already in `sphinx.sty`.

The list of active characters above can possibly be extended to include `+`, `:`, and  `"` because they are sanitized by `xcolor`. Already above I added the  colon `:`: for some reason `sphinxVerbatim` currently seems to have forgotten it from candidate characters for being made active (as they are not Pygmentized) to allow linebreaks. Possibly an oversight. Or perhaps some cautiousness for reasons I have forgotten.

One has to be careful with making characters active if the input contains macros where the characters may appear in some "string"-like argument (not typeset, but deciding what is done). And the ansi2latex produces such `\textcolor` and `\colorbox` usage as commented above. If no other LaTeX macros are added by mark-up we should be safe with above config.



> Currently the content of code cell outputs is passed through the filters `escape_latex` and `ansi2latex`, which are both provided by `nbconvert`:

Thanks for the links. Makes me think if Sphinx adopts some `\sphinxtextbackslash` mark-up it must be careful not to break `nbconvert` using `\textbackslash`.

mgeier commented:
Thanks, the new settings work very well, AFAICT. I've added them in  5803e465e08dfc934c30d06f37065fc2bf0dda4c

However, if you think that's a better solution, I could also create an additional custom filter to be applied after `nbconvert`'s `escape_latex`.

That way I could manually add some of the replacements from `sphinx.util.texescape.escape`.

I also tried using `sphinx.util.texescape.escape` directly as a filter (*instead of* `escape_latex`), but this replaced too many things. The problem was the literal `[` character which appears in all ANSI color sequences. And the ANSI decoding obviously doesn't work if `[` is replaced by `{[}`. But I guess all other replacements from `texescape.py` should be safe.
 
> So you can use `>` in a color name (not that I recommend it...)

I'm not planning to do that.
And I don't expect those names to change, so only the hyphen will be used as a special character.

> If no other LaTeX macros are added by mark-up we should be safe with above config.

All possible LaTeX markup should be in there: https://github.com/jupyter/nbconvert/blob/b0209dfd029da8797a144023b008198c3e36aa32/nbconvert/filters/ansi.py#L120-L168

jfbu commented:
Raw code such as `\textcolor[RGB]{255,0,0}{test}` would crash due to activation of `,` but this is already taken into account by the ansi2latex converter:

https://github.com/jupyter/nbconvert/blob/b0209dfd029da8797a144023b008198c3e36aa32/nbconvert/filters/ansi.py#L149-L154

For info, original LaTeX includes the comma `,` in the list of activated characters in verbatim output because `,, ` is a ligature in T1-encoded fonts. The default list is:

```
$ latexdef verbatim@nolig@list

\verbatim@nolig@list:
macro:->\do \`\do \<\do \>\do \,\do \'\do \-
```

The package `upquote` loaded automatically by Sphinx hacks into `\@noligs` in order to modify the way the single quotes are handled, thus not only are ligatures modified but the single quotes are rendered as straight quotes in output.

`sphinxVerbatim` wants to do more than preventing ligatures: it wants to allow linebreaks and for that it too needs to activate some characters but it assigns to them some definitions allowing the linebreaks (and they turn to also in passing prevent ligatures). As the processing by `fancyvrb` of `\verbatim@nolig@list` is done *after*  the Sphinx set-up, Sphinx has its own shorter `\sphinx@verbatim@nolig@list` (and I should have not used `@` letter in there) which locally replaces `\verbatim@nolig@list`.

Its default is only ``\do\` `` for following reasons:

- the `'`, `<`, `>`, and `-` should have been Pygmentized hence they were not expected to appear in contents, and then why activate them as it is always potentially dangerous to do that? And for the quote `'`, the Sphinx redefinition of `\PYGZsq` makes sure to also use `\textquotesingle` which gives a straight quote in PDF output.

- the comma `,` is not Pygmentized, hence it is handled in `\sphinxbreaksviaactive` to allow a linebreak after it, and this also prevents the T1 ``,, `` ligature,

- there only remained the `` ` ``, which is kept in the list to allow the "upquote" hack to work to let this become straight in PDF output.

In retrospect, Sphinx could have left the Pygmentized characters `'`, `<`, `>`, and `-` in the list: the rationale for not doing so was to avoid activating needlessly characters.

I am providing the above explanations as background info: what is to be done depends on whether you want or not to add to bare Verbatim treatement some potential linebreaks at specific characters.

Notice that `sphinxVerbatim` at any rate adds potential linebreaks at spaces. 

To limit its action to only that, you could redefine the `\sphinxbreaksbeforeactivelist`, `\sphinxbreaksafteractivelist` to be empty macros, and do in preamble

```
\makeatletter
\let\sphinx@verbatim@nolig@list\verbatim@nolig@list
\makeatother
\renewcommand*\sphinxbreaksbeforeactivelist{}
\renewcommand*\sphinxbreaksafteractivelist{}
```

in which case you end-up with almost same as `OriginalVerbatim` (aka `Verbatim`) with only the addition of linebreaks at spaces and consistent interline distances (and framing of course). And the dangers of the active characters in this context (active to avoid ligatures) are taken into account by the `ansi2latex` which uses `\detokenize`  to handle the comma, and also by `xcolor` (I am not sure `nbconvert` maintainers are aware of that) which sanitizes the hyphen `-` which may appear in color names as used by `ansi2latex`.

What is to be done depends very much on exactly what kind of contents can appear in these environments. For ANSI colors, it looks as if proposal of previous comment is ok, but if you want to be conservative you can use the  above which preserves the original `\verbatim@nolig@list` and does not activate anything thing else (hence no extra `\fvset{codes*=...}`).

jfbu commented:
(I posted previous comment prior to having read yours...)

jfbu commented:
> and do in preamble
> 
> ```
> \makeatletter
> \let\sphinx@verbatim@nolig@list\verbatim@nolig@list
> \makeatother
> \renewcommand*\sphinxbreaksbeforeactivelist{}
> \renewcommand*\sphinxbreaksafteractivelist{}
> ```
> 
> in which case you end-up with almost same as `OriginalVerbatim`

But this proposal is bad as it affects **all** `sphinxVerbatim`. Of course you could include it inside the "raw" directives only to limit it to the `sphinxVerbatim` for non-code contents.

On the other hand my earlier proposal

```
\makeatletter
\renewcommand*\sphinx@verbatim@nolig@list{\do\'\do\`}
\makeatother
\renewcommand*\sphinxbreaksbeforeactivelist{\do\<}
\renewcommand*\sphinxbreaksafteractivelist{\do\.\do\,\do\:\do\;\do\?\do\!\do\/\do\>\do\-}
\fvset{codes*=\sphinxbreaksattexescapedchars}
```

although affecting **all** `sphinxVerbatim` a priori does not alter the way the regular `sphinxVerbatim` containing Pygmentized contents do their job.

jfbu commented:
> Thanks, the new settings work very well, AFAICT. I've added them in [5803e46](https://github.com/spatialaudio/nbsphinx/commit/5803e465e08dfc934c30d06f37065fc2bf0dda4c)
> 
> However, if you think that's a better solution, I could also create an additional custom filter to be applied after `nbconvert`'s `escape_latex`.
> 
> That way I could manually add some of the replacements from `sphinx.util.texescape.escape`.
> 

Replacements of `sphinx.util.texescape.escape` are a legacy thing, I have not looked in detail at the rationale for each.

> I also tried using `sphinx.util.texescape.escape` directly as a filter (_instead of_ `escape_latex`), but this replaced too many things. The problem was the literal `[` character which appears in all ANSI color sequences. And the ANSI decoding obviously doesn't work if `[` is replaced by `{[}`. But I guess all other replacements from `texescape.py` should be safe.
> 

The `{[}` of `sphinx.util.texescape.escape` is probably related to prevent mis-interpretation in some contexts where a naked `[` would be interpreted as start of some optional argument.

For Unicode aspects, there is https://github.com/sphinx-doc/sphinx/pull/6750 

> > So you can use `>` in a color name (not that I recommend it...)
> 
> I'm not planning to do that.
> And I don't expect those names to change, so only the hyphen will be used as a special character.
> 
> > If no other LaTeX macros are added by mark-up we should be safe with above config.
> 
> All possible LaTeX markup should be in there: https://github.com/jupyter/nbconvert/blob/b0209dfd029da8797a144023b008198c3e36aa32/nbconvert/filters/ansi.py#L120-L168

OK, then things are fine I hope. The config I propose adds `:`  to activated characters, compared to current Sphinx which somehow overlooked it. (it is not Pygmentized, why did I not activate it? I was cautious I guess). You can extend even more list of activated characters (non letters, non digits), but do not do anything with braces `{`, `}`, and the backslash `\ `, and also don't activate the square brackets `[`, `]` because they appear in https://github.com/jupyter/nbconvert/blob/b0209dfd029da8797a144023b008198c3e36aa32/nbconvert/filters/ansi.py#L120-L168 syntax.

Thus linebreaks at `[`, `]`, and backslash are not possible. For backslash it was added recently at https://github.com/sphinx-doc/sphinx/pull/6799 but this works only because it is TeX-escaped. One can not do anything with the ``\ `` naked character which must conserve its meaning of staring TeX control sequences.

mgeier commented:
> what is to be done depends on whether you want or not to add to bare Verbatim treatement some potential linebreaks at specific characters.

TBH, I don't know.
This could be any output from some Python script, or from Bash, or whatever ...

Since it could be anything, I don't really know where line breaks make sense.
It would probably be best to try to mimic whatever `sphinxVerbatim` does, or what do you think?
I guess it would be strange to have different line breaks in code cell inputs (which use `docutils.nodes.literal_block` and therefore `sphinxVerbatim`) and code cell outputs, right?

> [...] but if you want to be conservative you can use the  above which preserves the original `\verbatim@nolig@list` and does not activate anything thing else (hence no extra `\fvset{codes*=...}`).

I don't know whether I want to be conservative or not.
But I think I prefer the less conservative option, because it seems to be closer to the behavior of input code cells.

> The config I propose adds `:`  to activated characters, compared to current Sphinx which somehow overlooked it.

OK, apparently this is consistent between input and output code cells, very good!

> Thus linebreaks at `[`, `]`, and backslash are not possible.

OK, that seems to be consistent with input code cells.

> For backslash it was added recently at https://github.com/sphinx-doc/sphinx/pull/6799 but this works only because it is TeX-escaped. One can not do anything with the ``\ `` naked character which must conserve its meaning of staring TeX control sequences.

I'm not sure if I completely understand, but backslashes are replaced by `escape_latex` and therefore there will never be a backslash on its own.

I've tried it with Sphinx' `2.0` branch, and lines are correctly broken after `\`:

![image](https://user-images.githubusercontent.com/705404/68682225-0f9cee00-0565-11ea-984a-16c6adbd67f1.png)

So I guess everything is fine now, right?

mgeier commented:
I've just found one difference: When I use `^`, this is replaced by `\^{}`, and no line break happens.
In the input, this is replaced by `\PYGZca{}`, and a line break happens afterwards.

jfbu commented:
>  I don't really know where line breaks make sense.
> It would probably be best to try to mimic whatever `sphinxVerbatim` does, or what do you think?

I tend to agree

> I've just found one difference: When I use `^`, this is replaced by `\^{}`, and no line break happens.
> In the input, this is replaced by `\PYGZca{}`, and a line break happens afterwards.

The problem is that nbconvert latex escaping is not like sphinx.util.texescape.escape the former uses `\^{}` the latter uses `\textasciicircum{}`, thus the action of `\sphinxbreaksattexescapechars` is not on right input. 

I will refactor in future to more easily adapt what `\sphinxbreaksattexescapechars`  does but here the simplest is to abuse that `\do` has correct meaning immediately after it so we can add it like below (very last line):


```
\makeatletter
\renewcommand*\sphinx@verbatim@nolig@list{\do\'\do\`}
\makeatother
\renewcommand*\sphinxbreaksbeforeactivelist{\do\<\do\"}
\renewcommand*\sphinxbreaksafteractivelist{\do\.\do\,\do\:\do\;\do\?\do\!\do\/\do\>\do\-}
\fvset{codes*=\sphinxbreaksattexescapedchars\do\^\^}
```

I also added the `\do\"` to `\sphinxbreaksbeforeactivelist` because normal sphinxVerbatim with Pygmentized contents does it with `\PYGZdq{}`.

The difference between `"` and `^` is that the former has to be made a TeX-active character but the latter was TeX-escaped to some macro `\^{}` and we only need redefine how that macro expands.

In both cases the above mimicks how they would behave if Pygmentized although now they are not Pygmentized.

And the config does not cancel or modify behaviour of sphinxVerbatim with Pygmentized contents because the macro `\sphinxbreaksatspecials` which handles Pygmentized contents is executed automatically in addition to the above. (I should have given it another name like `\sphinxbreaksatpygments`).


> I'm not sure if I completely understand, but backslashes are replaced by `escape_latex` and therefore there will never be a backslash on its own.


Yes, I was sort or speaking in a kind of ultimate generality which was near non-sense. That's the whole charm of my verbosity.

> 
> I've tried it with Sphinx' `2.0` branch, and lines are correctly broken after `\`:
> 
> ![image](https://user-images.githubusercontent.com/705404/68682225-0f9cee00-0565-11ea-984a-16c6adbd67f1.png)
> 
> So I guess everything is fine now, right?

Indeed! thanks for testing this with Sphinx dev repo.

jfbu commented:
edit: one would like to get the `'` to behave like in Pygmentized sphinxVerbatim. But there is a difficulty because the meaning of its active self is reset very late by Verbatim after Sphinx customizations of active characters. To work around that needs some non-trival LaTeX, sorry.


```
\makeatletter
\renewcommand*\sphinx@verbatim@nolig@list{\do\'\do\`}
\begingroup
\catcode`'=\active
\g@addto@macro\@noligs{\let'\PYGZsq}
\endgroup
\makeatother
\renewcommand*\sphinxbreaksbeforeactivelist{\do\<\do\"\do\'}
\renewcommand*\sphinxbreaksafteractivelist{\do\.\do\,\do\:\do\;\do\?\do\!\do\/\do\>\do\-}
\makeatletter
\fvset{codes*=\sphinxbreaksattexescapedchars\do\^\^}
\makeatother
```

jfbu commented:
The idea of previous trick in comment is that `\PYGZsq` is already re-defined by Sphinx to allow a linebreak at the time the `\let'\PYGZsq` will be done, so this avoids having to type the whole complicated `\discretionary...` thing.

jfbu commented:
This makes me think an even better way would be to generalize the `\let'\PYGZsq` trick to other characters so that really @mgeier `sphinxVerbatim` for plain-text inherits `sphinxVerbatim` Pygmentized behaviour *and* obeys user modifications of `\sphinxbreaksbeforelist/\sphinxbreaksafterlist` but I propose we call a truce here. And I was planning to have dinner...

jfbu commented:
memo: the modification to `\@noligs` in code above influences also inline-literals (`\sphinxupquote`) but there is no change in PDF output, instead of having active `'` a strict synonym of `\textquotesingle` (done by `upquote` package) it becomes synonym to `\PYGZsq` which is a macro expanding to `\textquotesingle` (defined in `sphinxhighlight.sty`) hence only very slight negligible overhead.

If the modification of `\@noligs` is to be strictly limited to `sphinxVerbatim`:

```
\makeatletter
\renewcommand*\sphinx@verbatim@nolig@list{\do\'\do\`}
\begingroup
\catcode`'=\active
\let\nbsphinx@noligs\@noligs
\g@addto@macro\nbsphinx@noligs{\let'\PYGZsq}
\endgroup
\makeatother
\renewcommand*\sphinxbreaksbeforeactivelist{\do\<\do\"\do\'}
\renewcommand*\sphinxbreaksafteractivelist{\do\.\do\,\do\:\do\;\do\?\do\!\do\/\do\>\do\-}
\makeatletter
\fvset{codes*=\sphinxbreaksattexescapedchars\do\^\^\let\@noligs\nbsphinx@noligs}
\makeatother
```

Sorry. The problem is that a priori we don't want our modifications to perturbate `\sphinxupquote`.

mgeier commented:
Wow, thanks for all this TeX magic!

I've squashed the previous commits into 791c6eae8d73f260a33ccbc9d5a321131d7f681f and added you as a co-author, because I clearly wouldn't have been able to come up with any of this.

jfbu commented:
@mgeier thanks for the honor! I suddenly dearly hope nothing catastrophic will result from merging this (my insolvency is already in place, in case some litigation ensues :) )

I will be careful if something evolves at Sphinx side it does not break this...

mgeier commented:
Don't worry, if something breaks we can fix it again!

And to avoid litigation, we have this:

https://github.com/spatialaudio/nbsphinx/blob/5faf5bb84110f01591f24f09df2315e213181263/LICENSE#L13-L19
